tests/FSharp.Compiler.ComponentTests/Language/DiscriminatedUnionTests.fs

```diff
@@ -0,0 +1,160 @@
+    let ``Is* discriminated union properties with backticks are visible, proper values are returned`` () =
+        Fsx """
+type Foo = | Foo of string | ``Mars Bar``
+let foo = Foo.Foo "hi"
+if not foo.IsFoo then failwith "Should be Foo"
+if foo.``IsMars Bar`` then failwith "Should not be ``Mars Bar``"
+
+let marsbar = ``Mars Bar``
+if marsbar.IsFoo then failwith "Should not be Foo"
+if not marsbar.``IsMars Bar`` then failwith "Should be ``Mars Bar``"
+        """
+        |> withLangVersionPreview
+        |> compileExeAndRun
+        |> shouldSucceed
+
+    [<FSharp.Test.FactForNETCOREAPP>]
+    let ``Is* discriminated union properties are visible, proper values are returned in recursive namespace, before the definition`` () =
+        FSharp """
+namespace rec Hello
+
+module Main =
+    [<EntryPoint>]
+    let main _ =
+        let foo = Foo.Foo "hi"
+        if not foo.IsFoo then failwith "Should be Foo"
+        if foo.IsBar then failwith "Should not be Bar"
+        0
+
+[<Struct>]
+type Foo =
+    | Foo of string
+    | Bar
+        """
+        |> withLangVersionPreview
+        |> compileExeAndRun
+        |> shouldSucceed
+
+
+    [<FSharp.Test.FactForNETCOREAPP>]
+    let ``Is* discriminated union properties are visible, proper values are returned in recursive namespace, in SRTP`` () =
+        FSharp """
+namespace Hello
+
+[<Struct>]
+type Foo =
+    | Foo of string
+    | Bar
+
+module Main =
+
+    let inline (|HasIsFoo|) x = fun () -> (^a : (member IsFoo: bool) x)
+    let inline (|HasIsBar|) x = fun () -> (^a : (member IsBar: bool) x)
+    let getIsFooIsBar (HasIsFoo isFoo & HasIsBar isBar) = (isFoo(), isBar())
+
+    [<EntryPoint>]
+    let main _ =
+        let foo = Foo.Foo "hi"
+        let (isFoo, isBar) = getIsFooIsBar foo
+        if not isFoo then failwith "Should be Foo"
+        if isBar then failwith "Should not be Bar"
+        0
+        """
+        |> withLangVersionPreview
+        |> compileExeAndRun
+        |> shouldSucceed
+
+    [<FSharp.Test.FactForNETCOREAPP>]
+    let ``Is* discriminated union properties are unavailable with DefaultAugmentation(false)`` () =
+        Fsx """
+[<DefaultAugmentation(false)>]
+type Foo = | Foo of string | Bar
+let foo = Foo.Foo "hi"
+let isFoo = foo.IsFoo
+        """
+        |> withLangVersionPreview
+        |> typecheck
+        |> shouldFail
+        |> withErrorMessage "The type 'Foo' does not define the field, constructor or member 'IsFoo'. Maybe you want one of the following:
+   Foo"
+
+    [<FSharp.Test.FactForNETCOREAPP>]
+    let ``Is* discriminated union properties are unavailable on voption`` () =
+        Fsx """
+let x = (ValueSome 1).IsSome
+let y = ValueOption<int>.None.IsValueNone
+        """
+        |> withLangVersionPreview
+        |> typecheck
+        |> shouldFail
+        |> withErrorMessage "The type 'ValueOption<_>' does not define the field, constructor or member 'IsValueNone'. Maybe you want one of the following:
```

Comment:
@kerams :

I am having a problem with this test in a different PR and I cannot get my head around it - how is it ensured, that the generation is prevented for ValueOption?
Afterall, it does NOT have DefaultAugmentation(false).

What am I missing?
I tried to scan this PR's diff for any kind of logical condition that would skip IsValueNone and IsValueSome for ValueOption, but I do not see it.

Reply:
While the case testers exist, they aren't exposed in prim-types.fsi.

Reply:
And this PR does not add them automatically to anything in prim-types unless specified, or how does it work?
i.e. what condition is in place to make sure that ValueOption does NOT get them and a user-defined DU gets them?

Reply:
Yes, if there's a signature file, case testers need to be exposed manually. The RFC mentions it briefly.

Reply:
Thanks, got it.

For some reason, I have it keep coming in the nullness PR (e.g. ValueOption does come up with the IsValueNone tester), not sure what is so different over there.